dhis2/dhis2-core
# feat: email-based authentication codes for MFA/2FA

## Summary
This adds 2FA login with a code sent via email as a new feature.

NOTE: This PR removes the enforced 2FA enrollment with TOTP functionality and does not address this for email 2FA either. The requirements for enforced enrollment needs to be addressed in a separate JIRA ticket. Enforced enrollment for TOTP was never used, so it's safe to remove.

TODO:
* Add rate limiting to email 2FA sending
* Do sending of 2FA email async so we don't block UI
* Enforced enrollment

### Automatic tests
* TwoFactorControllerTest
* UserServiceTest
* AuthenticationControllerTest
* LoginTest (E2E)

### Manual tests
* Enable 2FA email in system settings: POST api/42/systemSettings/email2FAEnabled?value=true
* Set up a working email SMTP server in settings
* Verify your email: POST api/account/sendEmailVerification (email with code will be sent to your verified email)
* Enroll in email 2FA: POST api/2fa/enrollEmail2FA
* Enable email 2FA with code sent to your email: POST api/2fa/enable with {code:CODE_SENT_TO_EMAIL}
* Try to login, code should be sent to email
* Now login should show 2FA code input, enter the code and submit
* Observe you are logged in

### Jira:
[DHIS2-13334](https://dhis2.atlassian.net/browse/DHIS2-13334)


[DHIS2-13334]: https://dhis2.atlassian.net/browse/DHIS2-13334?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
